src/persist-client/src/read.rs

```diff
@@ -0,0 +1,224 @@
+pub struct SnapshotIter<K, V, T, D>(PhantomData<(K, V, T, D)>);
+
+impl<K, V, T, D> SnapshotIter<K, V, T, D>
+where
+    K: Debug + Codec,
+    V: Debug + Codec,
+    T: Timestamp + Lattice + Codec64,
+    D: Semigroup + Codec64,
+{
+    /// The frontier at which we're outputting the contents of the shard.
+    pub fn as_of(&self) -> &Antichain<T> {
+        todo!()
+    }
+
+    /// Attempt to pull out the next values of this iterator.
+    ///
+    /// An empty vector is returned if this iterator is exhausted.
+    pub async fn poll_next(
+        &mut self,
+        timeout: Duration,
+    ) -> Result<Vec<((Result<K, String>, Result<V, String>), T, D)>, LocationError> {
+        todo!("{:?}", timeout)
+    }
+}
+
+/// Data and progress events of a shard subscription.
+///
+/// TODO: Unify this with [timely::dataflow::operators::to_stream::Event] or
+/// [timely::dataflow::operators::capture::event::Event].
+pub enum ListenEvent<K, V, T, D> {
+    /// Progress of the shard.
+    Progress(Antichain<T>),
+    /// Data of the shard.
+    Updates(Vec<((Result<K, String>, Result<V, String>), T, D)>),
+}
+
+/// An ongoing subscription of updates to a shard.
+pub struct Listen<K, V, T, D>(PhantomData<(K, V, T, D)>);
+
+impl<K, V, T, D> Listen<K, V, T, D>
+where
+    K: Debug + Codec,
+    V: Debug + Codec,
+    T: Timestamp + Lattice + Codec64,
+    D: Semigroup + Codec64,
+{
+    /// Attempt to pull out the next values of this subscription.
+    pub async fn poll_next(
+        &mut self,
+        timeout: Duration,
+    ) -> Result<ListenEvent<K, V, T, D>, LocationError> {
+        todo!("{:?}", timeout)
+    }
+}
+
+/// A "capability" granting the ability to read the state of some shard at times
+/// greater or equal to `self.since()`.
+pub struct ReadHandle<K, V, T, D>
+where
+    T: Timestamp + Lattice + Codec64,
+{
+    _phantom: PhantomData<(K, V, T, D)>,
+}
+
+impl<K, V, T, D> ReadHandle<K, V, T, D>
+where
+    K: Debug + Codec,
+    V: Debug + Codec,
+    T: Timestamp + Lattice + Codec64,
+    D: Semigroup + Codec64,
+{
+    /// This handle's `since` frontier.
+    ///
+    /// This will always be greater or equal to the shard-global `since`.
+    pub fn since(&self) -> &Antichain<T> {
+        todo!()
+    }
+
+    /// Forwards the since frontier of this handle, giving up the ability to
+    /// read at times not greater or equal to `new_since`.
+    ///
+    /// This may trigger (asynchronous) compaction and consolidation in the
+    /// system. A `new_upper` of the empty antichain "finishes" this shard,
+    /// promising that no more data will ever be read by this handle.
+    ///
+    /// The clunky two-level Result is to enable more obvious error handling in
+    /// the caller. See <http://sled.rs/errors.html> for details.
+    pub async fn downgrade_since(
+        &self,
+        timeout: Duration,
+        new_since: Antichain<T>,
+    ) -> Result<Result<(), InvalidUsage>, LocationError> {
```

Comment:
I'm probably confused, but I would think that the error types should be swapped here. This would allow a caller to bubble up the `InvalidUsage` error using `?`. These usage errors always need to bubble up and stop the system because they represent bugs, I think. And then only the "real" `LocationError` remains. Could also be that I misunderstood the _sled.rs_ piece and have to go back and read it again.

Reply:
good point. my reading of the sled errors post was the same as yours and that's how I started. I swapped to this because it made error handling far easier internally in the prototype (we got to use ? far more often) and the brief bit of thinking I did about it made me think the caller doesn't actually care about the order, just that they're two separate results with different error types. but I consider this unresolved, especially until we've got a bit of usage of this api under our belt, so added this as a top level TODO